File: Java/src/main/java/QuestionBank/Solution201_300/Solution235.java
package QuestionBank.Solution201_300;

import tool.TreeNode;

/**
 * @author windows
 */
public class Solution235 {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        TreeNode ancestor = root;
        while (true){
            if (ancestor.val > p.val && ancestor.val > q.val){
                ancestor = ancestor.left;
            }else if (ancestor.val < p.val && ancestor.val < q.val){
                ancestor = ancestor.right;
            }else {
                break;
            }
        }
        return ancestor;
    }


    public static void main(String[] args) {
        Solution235 solution235 = new Solution235();
    }
}

File: C++/src/Solution201_300/Solution235.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution235.h"
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;
#include <cstring>
using namespace std;

//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};



TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
    if (root == nullptr || root == p || root == q) {
      return root;
    }
    if (p->val < root->val && q->val < root->val){
      return lowestCommonAncestor(root->left, p, q);
    }
    if (p->val > root->val && q->val > root->val){
      return lowestCommonAncestor(root->right,p, q);
    }
    return root;
}


